### ISSUE TITLE
Unexpected Egress Failures on TCP Route

### ISSUE THREAD

**Issue by smy2748** [aws/aws-app-mesh-roadmap#82]
Hello,

While setting up a service in AppMesh I encountered something that seems strange to me.  Our service is set up to call a postgres database (port 5432) outside of the appmesh, so we enabled all egress traffic.  When the virtual route is set as having an http protocol, the container is able to connect to the database.  The strange thing (to me at least) is that when the virtual route is set to have a tcp protocol (and that is the only change made to cause this problem), suddenly the container can no longer connect to the server (and also fails sending metrics out to a different external service as well).  

To me it seems odd that the egress traffic ALLOW_ALL setting stops working when the ingress protocol for the virtual route is changed.  Do you know if this is expected behavior with tcp routes?

**Comment by lavignes** [comment]
Hey @smy2748 
All things being equal, the behavior for egress tcp and http traffic should be the same. Can you clarify what you mean when you say "virtual route", do you mean Virtual Router or Route?

Regardless of that. Can you give us an example of your mesh configuration so we can reproduce your issue? There could certainly be a bug on our end.

**Comment by shubharao** [comment]
@smy2748 can you help us debug this issue by providing some more details on this?

**Comment by smy2748** [comment]
Yeah sure.  Sorry about the delay.  We have a bunch of gRPC services we're running on fargate and appmesh.  During setup I noticed that one of the services kept failing into an unhealthy state due to being unable to connect to our database (hosted in EC2).  I then checked that configuration against the one done in our staging mesh and found that the only difference was the protocol setting for the virtual route.  When I switched the virtual route over to using http the container was once again able to connect to the database.  

Configuration for successful service:
Egress Filter: Allow External Traffic
service listener protocol: http
service route protocol: http
service node protocol: tcp

Configuration for failing service:
Egress Filter: Allow External Traffic
service listener protocol: http
service route protocol: tcp
service node protocol: tcp

The big thing was that it was counterintuitive to debug, since I didn't expect the ingress routing protocol settings to affect egress traffic from the container.

**Comment by kiranmeduri** [comment]
@smy2748 I am unable to repro the above use-case but documented below the behavior with different possibilities w.r.t allow_all filter and protocol combinations. You are right that ingress routing for a node is used by the callers of this node and should not impact application's egress traffic. I am afraid there is something else that is causing the problem. If you can provide virtual-node/virtual-router spec for _service_ along with a system diagram it will help. Additionally **sanitized** envoy logs/config-dump may help to debug further.

One additional thing is that if you are updating virtual-node/virtual-router protocol on the same port then Envoy seems to be rejecting an update. 

**tl;dr**
Virtual-router contain routes that target one or more virtual-nodes. Protocols supported by virtual-router listener should match virtual-node listener. If they do not match behavior is un-predictable. Read further details below on how the behavior changes.

|         | **vn-http** | **vn-tcp**  |
|---------|---------|---------|
| **vr-http** | success | fail    |
| **vr-tcp**  | fail    | success |

### Details 

Code: https://github.com/kiranmeduri/aws-app-mesh-examples/tree/bug-82/walkthroughs/bug-82

I tried with setup ` FrontApp -> ColorApp -> Postgres DB (RDS)` where FrontApp listens on port 8080, ColorApp listens on port 9080, and Postgres DB listens on port 5432. Following are the 4 combinations w.r.t color-app

1. ColorApp virtual-node with HTTP/9080, virtual-router with HTTP/9810
   1. Result: **SUCCESS** `Front -> Color -> DB`
   2. Egress listener of front app is correctly configured with HTTP filter as shown below;
   ```
    ...
    "route_config": {
      "name": "rds_egress_http_9810",
      "virtual_hosts": [
        {
          "name": "color.bug-82.mesh.local",
          "domains": [
            "color.bug-82.mesh.local",
            "color.bug-82.mesh.local:9810"
          ],
          "routes": [
            {
              "match": {
                "prefix": "/"
              },
              "route": {
                "weighted_clusters": {
                  "clusters": [
                    {
                      "name": "cds_egress_bug-82_bug-82-color-node_http_9080",
                      "weight": 50
                    },
                    {
                      "name": "cds_egress_bug-82_bug-82-color-v2-node_http_9080",
                      "weight": 50
                    }
                  ],
                  "total_weight": 100
                }
              }
            }
          ]
        },
        {
          "name": "mesh_allow_all",
          "domains": [
            "*"
          ],
          "routes": [
            {
              "match": {
                "prefix": "/"
              },
              "route": {
                "cluster": "cds_egress_bug-82_mesh-allow-all"
              }
            }
          ]
        }
      ]
    }
   ...
   ```
2. ColorApp virtual-node with HTTP/9080, virtual-router with TCP/9811
   1. Result: **FAIL** `Front -> Color -> DB`
   2. Reason:  Routes to color app are not properly configured when virtual-router's listener protocol and target virtual-node listener's protocol do not match. In this case the egress listener of front app is missing a filter for 9811. 
3. ColorApp virtual-node with TCP/9080, virtual-router with HTTP/9812
   1. Result: **FAIL** `Front -> Color -> DB`
   2. Reason: Routes to color app are not properly configured when virtual-router's listener protocol and target virtual-node listener's protocol do not match. In this case the egress listener of front app is using HTTP connection manager as expected due to HTTP virtual-router, but corresponding route_config is wrong. This is **BUG** that needs to be fixed on App Mesh side.
   ```
    ...
    {
      "filter_chain_match": {
        "prefix_ranges": [
          {
            "address_prefix": "0.0.0.0",
            "prefix_len": 0
          }
        ],
        "destination_port": 9812
      },
      "filters": [
        {
          "name": "envoy.http_connection_manager",
          "config": {
            "tracing": {
              "operation_name": "EGRESS"
            },
            "http_filters": [
              {
                "name": "envoy.router"
              }
            ],
            "rds": {
              "route_config_name": "rds_egress_http_9812",
              "config_source": {
                "ads": {}
              }
            },
            "stat_prefix": "egress"
          }
        }
      ]
    }
    ...
    "route_config": {
      "name": "rds_egress_http_9812",
      "virtual_hosts": [
        {
          "name": "color.bug-82.mesh.local",
          "domains": [
            "color.bug-82.mesh.local",
            "color.bug-82.mesh.local:9812"
          ]
          !!!! MISSING ROUTE and CLUSTER target !!!!
        },
        {
          "name": "mesh_allow_all",
          "domains": [
            "*"
          ],
          "routes": [
            {
              "match": {
                "prefix": "/"
              },
              "route": {
                "cluster": "cds_egress_bug-82_mesh-allow-all"
              }
            }
          ]
        }
      ]
    }
    ...
   ```

4. ColorApp virtual-node with TCP/9080, virtual-router with TCP/9813
   1. Result: **SUCCESS** `Front -> Color -> DB`
   2. Egress listener of front app is correctly configured with TCP filter as shown below;
   ```
   ...
    {
      "filter_chain_match": {
        "prefix_ranges": [
          {
            "address_prefix": "0.0.0.0",
            "prefix_len": 0
          }
        ],
        "destination_port": 9813
      },
      "filters": [
        {
          "name": "envoy.tcp_proxy",
          "config": {
            "stat_prefix": "egress",
            "weighted_clusters": {
              "clusters": [
                {
                  "name": "cds_egress_bug-82_bug-82-color-node_tcp_9080",
                  "weight": 50
                },
                {
                  "name": "cds_egress_bug-82_bug-82-color-v2-node_tcp_9080",
                  "weight": 50
                }
              ]
            }
          }
        }
      ]
    }
   ...
   ```

**Comment by bcelenza** [comment]
@smy2748 Hey there, curious if the information above has helped resolve this issue? Let us know if not and we can dig in further.

**Comment by kiranmeduri** [comment]
Closing this, please re-open if you need more information. Thanks
